### PR TITLE
Configuration flags on the commandline

### PR DESCRIPTION
**Pull request by karanankit01**
Set the configuration flags input from commandline as stated in issue#[110](https://github.com/Koihik/LuaFormatter/issues/110)

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -9,6 +9,142 @@
...
+    args::Flag trueSingleDouble(SingleDouble, "single to double quote", "Transform string literals to use double quote", {"single-quote-to-double-quote"});
+    args::Flag falseSingleDouble(SingleDouble, "single to double quote", "Do not transform string literals to use double quote", {"no-single-quote-to-double-quote"});
+    if (trueSingleDouble) { config.ArgmapBool["single_quote_to_double_quote"] = true; }
+    if (falseSingleDouble) { config.ArgmapBool["single_quote_to_double_quote"] = false; }
+
+}
+
 int main(int argc, const char* argv[]) {
     args::ArgumentParser parser("Reformats your Lua source code.", "");
     args::HelpFlag help(parser, "help", "Display this help menu", {'h', "help"});
```
Comment (tammela) [inline src/main.cpp:150]: Move this `args::` block to the function that process the other arguments

File: src/Config.h [hunk 2]
```diff
@@ -10,6 +10,12 @@ class Config {
     void readFromFile(const string &file);
     void dumpCurrent(ofstream &fout);
     void dumpCurrent(ostream &out);
+    map<string, int> ArgmapInt;
```
Comment (tammela) [inline src/Config.h:13]: We can use a map with type `String -> Any` in plain C++ 17 and unify the 3 implementations.
See: https://en.cppreference.com/w/cpp/utility/any

File: src/main.cpp [hunk 3]
```diff
@@ -9,6 +9,142 @@
 
 namespace fs = filesystem;
 
+void ArgConfig( args::ArgumentParser &parser) {
+
+    Config config;
```
Comment (tammela) [inline src/main.cpp:14]: This is wrong, take `config` as a reference from the `config` declared in `main()`.

File: src/main.cpp [hunk 4]
```diff
@@ -9,6 +9,142 @@
...
+    if (tablesep) { config.ArgmapChar["table_sep"] = args::get(tablesep); }
+
+    if (columnlimit) { config.ArgmapInt["column_limit"] = args::get(columnlimit); }
+    if (indentwidth) { config.ArgmapInt["indent_width"] = args::get(indentwidth); }
+    if (tabwidth) { config.ArgmapInt["tab_width"] = args::get(tabwidth); }
+    if (continuationindentwidth) { config.ArgmapInt["continuation_indent_width"] = args::get(continuationindentwidth); }
+    if (spacesbeforecall) { config.ArgmapInt["spaces_before_call"] = args::get(spacesbeforecall); }
+    if (columntablelimit) { config.ArgmapInt["column_table_limit"] = args::get(columntablelimit); }
+
+    args::Group usetab(parser, "This group is all exclusive:", args::Group::Validators::Xor);
```
Comment (tammela) [inline src/main.cpp:32]: Please review this in the entire code.
Change `usetab` to `optusetab`.
Change the options to `usetab` for true and `nousetab` for false.

File: src/main.cpp [hunk 5]
```diff
@@ -9,6 +9,142 @@
...
+    args::Flag trueusetab(usetab, "Tab indentation", "Use tab for indentation", {"use-tab"});
+    args::Flag falseusetab(usetab, "Tab indentation", "Do not use tab for indentation", {"no-use-tab"});
+    if (trueusetab) { config.ArgmapBool["use_tab"] = true; }
+    if (falseusetab) { config.ArgmapBool["use_tab"] = false; }
+
+    args::Group controlblock(parser, "This group is all exclusive:", args::Group::Validators::Xor);
+    args::Flag truecontrolblock(controlblock, "Control block", "keep block in one line", {"keep-simple-control-block-one-line"});
+    args::Flag falsecontrolblock(controlblock, "Control block", "Do not keep block in one line", {"no-keep-simple-control-block-one-line"});
+    if (truecontrolblock) { config.ArgmapBool["keep_simple_control_block_one_line"] = true; }
+    if (falsecontrolblock) { config.ArgmapBool["keep_simple_control_block_one_line"] = false; }
```
Comment (tammela) [inline src/main.cpp:42]: Please review this in the entire code.

`else if`

File: src/main.cpp [hunk 6]
```diff
@@ -9,6 +9,142 @@
...
+    Config config;
+
+    args::ValueFlag<int> columnlimit(parser, "column limit", "Column limit of one line", {"column-limit"});
+    args::ValueFlag<int> indentwidth(parser, "indentation width", "Number of spaces used for indentation", {"indent-width"});
+    args::ValueFlag<int> tabwidth(parser, "tab width", "Number of spaces used per tab", {"tab-width"});
+    args::ValueFlag<int> continuationindentwidth(parser, "Continuation indentation width", "Indent width for continuations line", {"continuation-indent-width"});
+    args::ValueFlag<int> spacesbeforecall(parser, "spaces before call", "Space on function calls", {"spaces-before-call"});
+    args::ValueFlag<int> columntablelimit(parser, "column table limit", "Column limit of each line of a table", {"column-table-limit"});
+    args::ValueFlag<char> tablesep(parser, "table separator", "Character to separate table fields", {"table-sep"});
+    if (tablesep) { config.ArgmapChar["table_sep"] = args::get(tablesep); }
```
Comment (tammela) [inline src/main.cpp:23]: These checks are wrong and will have no effect here, you gotta do those after we call `parser.ParseCLI(argc, argv);`.
So split the function into 2.
One that setups the arguments.
Another that reads the arguments after the library parsed them.

File: src/main.cpp [hunk 7]
```diff
@@ -9,6 +9,142 @@
 
 namespace fs = filesystem;
 
+void ArgConfig( args::ArgumentParser &parser) {
```
Comment (tammela) [inline src/main.cpp:12]: Always use `clang-format` with the project's style before posting the PR.

File: src/main.cpp [hunk 8]
```diff
@@ -67,7 +321,7 @@ int main(int argc, const char* argv[]) {
     }
 #endif
 
-    if (configFileName.empty()) {
+    if (configFileName.empty() && config.Argmap.empty()) {
```
Comment (tammela) [inline src/main.cpp:324]:
```suggestion
    if (configFileName.empty()) {
```
You are forcing to use the configuration via command line options only when there is a configuration file.

File: src/main.cpp [hunk 9]
```diff
@@ -80,6 +334,10 @@ int main(int argc, const char* argv[]) {
         return 1;
     }
 
+    if (!config.Argmap.empty()) {
```
Comment (tammela) [inline src/main.cpp:337]: Move this new block after the `use_default` label.

File: src/Config.h [hunk 10]
```diff
@@ -10,6 +11,8 @@ class Config {
     void readFromFile(const string &file);
     void dumpCurrent(ofstream &fout);
     void dumpCurrent(ostream &out);
+    map<string, any> Argmap;
```
Comment (Koihik) [inline src/Config.h:14]: This should be in lowercase.

File: src/Config.cpp [hunk 11]
```diff
@@ -77,6 +78,30 @@ void Config::readFromFile(const string &file) {
...
 }
 
+void Config::readFromMap(map<string, any> &mp) {
+    bool given = false;
+    string CTL = "column_table_limit";
+    string CL = "column_limit";
+    for (auto kv : mp) {
+        auto key = kv.first;
+        if (node_[key]) {
+            if (*((kv.second).type().name()) == 'i') {
```
Comment (Koihik) [inline src/Config.cpp:88]: Looks strange. Why not just write `kv.second.type().name() == "int"`

Comment (karanankit01) [reply src/Config.cpp:88]: If we not use *, it shows 
`ISO C++ forbids comparison between pointer and character`
And, `*((kv.second).type().name())` it gives character as 'i' for int, 'c' for character, 'b' for bool so on.

File: src/Config.h (hunk 10 above)
Comment (karanankit01) [reply src/Config.h:14]: ok, I will change `Argmap` to `argmap`.

File: src/Config.cpp (hunk 11 above)
Comment (Koihik) [reply src/Config.cpp:88]: Convert to std::string or use strcmp

Comment (karanankit01) [reply src/Config.cpp]: Done.

File: src/Config.cpp [hunk 12]
```diff
@@ -77,6 +79,30 @@ void Config::readFromFile(const string &file) {
...
 }
 
+void Config::readFromMap(map<string, any> &mp) {
+    bool given = false;
+    string CTL = "column_table_limit";
+    string CL = "column_limit";
+    for (auto kv : mp) {
+        auto key = kv.first;
+        if (node_[key]) {
+            if (strcmp(kv.second.type().name(), "i") == 0) {
```
Comment (Koihik) [inline src/Config.cpp:89]: `strcmp(kv.second.type().name(), "int")`

Comment (karanankit01) [reply src/Config.cpp:89]: it returns "i" not  "int"

Comment (Koihik) [reply src/Config.cpp:89]: OK

**Pull request merged**